File: leetcode/2551/c++/solution.cpp
/*
    2551. Put Marbles in Bags
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    long long putMarbles(vector<int>& weights, int k) {
        long long output = 0;
        if (k > 1) {
            std::priority_queue<int> max_w, min_w;
            for (size_t i = 1; i < weights.size(); ++i) {
                min_w.push(weights[i - 1] + weights[i]);
                max_w.push(-(weights[i - 1] + weights[i]));
                if (min_w.size() > k - 1) {
                    min_w.pop();
                    max_w.pop();
                }
            }
            long long min_score = weights[0] + weights[weights.size() - 1];
            long long max_score = weights[0] + weights[weights.size() - 1];
            while (!min_w.empty()) {
                min_score += min_w.top();
                min_w.pop();
                max_score += -max_w.top();
                max_w.pop();
            }
            output = max_score - min_score;
        }
        return output;
    }
};

File: leetcode/2551/rust/solution.rs
/*
    2551. Put Marbles in Bags
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn put_marbles(weights: Vec<i32>, k: i32) -> i64 {
        let mut output: i64 = 0;
        let k: usize = k as usize - 1;
        if (k > 0 && weights.len() - 1 > k) {
            let mut weights_mut: Vec<i32> = Vec::with_capacity(weights.len());
            for i in 0..(weights.len() - 1) {
                weights_mut.push(weights[i] + weights[i + 1]);
            }
            weights_mut.select_nth_unstable(k);
            for i in 0..(k as usize) {
                output -= weights_mut[i] as i64;
            }
            let n = weights_mut.len() - k - 1;
            weights_mut.select_nth_unstable(n);
            for i in 0..(k as usize) {
                output += weights_mut[weights_mut.len() - i - 1] as i64;
            } //*/
        }
        output
    }
}
